Before you turn this problem in, make sure everything runs as expected. First, **restart the kernel** (in the menubar, select Kernel$\rightarrow$Restart) and then **run all cells** (in the menubar, select Cell$\rightarrow$Run All).

Make sure you fill in any place that says `YOUR CODE HERE` or "YOUR ANSWER HERE", as well as your name and collaborators below:

In [ ]:
NAME = "Michael Tran"
COLLABORATORS = "Junnuo Zhu, Ryan Whitford, Jaden Mak"

# Physics C170M/270M Lab 2: Naive Bayes 

In this lab, you will
- Incorporate the 8 steps of machine learning 
- Calculate a small Naive Bayes problem by hand
- Code a Gaussian Naive Bayes model to analyze the iris dataset. 
- Use scikit-learn and Naive Bayes models to make predictions
- Application of Naive Bayes in Astrophysics: classify stars at the center of our Galaxy

Reminder: save and checkpoint often!

Lab Created by: Tuan Do & Bernie Boscoe

Last updated by: Tuan Do

---

# Naive Bayes exercise by hand, with machine learning language
Bayes theorem:
$P(A \mid B) = \frac{P(A, B)}{P(B)} = \frac{P(B\mid A) \times P (A)}{P(B)}$
Basically, we are trying to find probability of event A, given the event B is true. Event B is also termed as evidence.
P(A) is the prior of A (the prior probability, i.e. Probability of event before evidence is seen). The evidence is an attribute value of an unknown instance(here, it is event B).
P(A|B) is a posteriori probability of B, i.e. probability of event after evidence is seen.

Let $(x_{1},x_{2},…,x_{p})$ be a feature vector and y be the class label corresponding to this feature vector.
Applying Bayes’ theorem,

$$P(y \mid X) = \frac{P(X, y)}{P(X)} = \frac{P(X\mid y) \times P (y)}{P(X)}$$

where X is given as $ X=(x_{1},x_{2},…,x_{p})$. 
By substituting for X and expanding using the chain rule we get,

$$P(y \mid x_{1}, x_{2},..., x_{p}) = \frac{P(x_{1}, x_{2},..., x_{p}, y)}{P(x_{1}, x_{2},..., x_{p})} = \frac{P(x_{1}, x_{2},..., x_{p}\mid y) \times P (y)}{P(x_{1}, x_{2},..., x_{p})}$$

Since $ (x_{1},x_{2},…,x_{p})$ are independent of each other,

$$P(y \mid x_{1}, x_{2},..., x_{p}) = \frac{P (y) \times \prod_{i=1}^{p} P(x_{i} \mid y)}{\prod_{i=1}^{p} P(x_{i})}$$

For all entries in the dataset, the denominator does not change, it remains constant. Therefore, the denominator can be removed and a proportionality can be introduced:

$$P(y \mid x_{1}, x_{2},..., x_{p}) \propto P (y) \times \prod_{i=1}^{p} P(x_{i} \mid y)$$

In our scikit learn exercise, the response variable (y) has only two outcomes, binary (e.g., yes or no / positive or negative). There could be cases where the classification could be multivariate.

To complete the specification of our classifier, we adopt the MAP (Maximum A Posteriori) decision rule, which assigns the label to the class with the highest posterior.

$$\hat{y} = p(X, y) = p(y, x_{1}, x_{2},..., x_{p}) = \operatorname*{argmax}_{k \in \{1,2, ...,K\}} P (y) \times \prod_{i=1}^{p} P(x_{i} \mid y)$$

We calculate probability for all ‘K’ classes using the above function and take one with the maximum value to classify a new point belongs to that class.




Let's look at a naive bayes by hand example, using data science jargon. The data we will use is various aspects of weather conditions and a binary outcome, whether or not one plays golf.

The dataset is divided into two parts, namely, feature matrix and the response vector.

The feature matrix contains all the vectors(rows) of dataset in which each vector consists of the value of dependent features. In our dataset, features are ‘Temperature’, ‘Humidity’ and ‘Weather’.
Response vector contains the value of class variable(prediction or output) for each row of feature matrix. In our dataset, the class variable name is ‘Play golf’. Here is a list of the categorical features and the values they take, as well as the binary predicted value.

|Temperature         | Weather         | Humidity   | Play Golf |
| ------------- |-------------| ------|--------|
| Hot   | Sunny | High |Yes|
| Cold     | Rainy      |  Low |No

Let's look at ten events and outcomes:

|Temperature         | Weather         | Humidity   | Play Golf |
| ------------- |-------------| ------|--------|
| Hot   | Sunny | High |Yes|
| Hot    | Rainy      |  High |Yes|
| Hot   | Rainy | High |Yes|
| Cold     | Sunny |  Low |No|
| Hot   | Sunny | Low |No|
| Cold     | Rainy|  Low |No|
| Cold   | Sunny | High |Yes|
| Cold     | Rainy |  Low |No|
| Cold  | Sunny | High |No|
| Cold     | Rainy |  Low |Yes|

Just to clear, an example of a feature vector and corresponding class variable can be: (refer to 1st row of dataset)

X = (Hot, Sunny , High)

y = Yes

So basically, P(y|X) here means, the probability of “playing golf” given that the weather conditions are “Temperature is hot”, "sunny", and “high humidity”.

# Naive assumption

Now, its time to put a naive assumption to the Bayes’ theorem, which is, independence among the features. So now, we split evidence into the independent parts.

Now, if any two events A and B are independent, then,

P(A,B) = P(A)P(B)

So, finally, we are left with the task of calculating $P(y)$ and $P(x_i | y).$

Please note that $P(y)$ is also called the class probability and $P(x_i | y)$ is called the conditional probability.

The different naive Bayes classifiers differ mainly by the assumptions they make regarding the distribution of $P(x_i | y)$.

Let us try to apply the above formula manually on our weather dataset. For this, we need to do some precomputations on our dataset.

We need to find $P(x_i | y_j$) for each $x_i$ in $X$ and $y_j$ in y. 



Fill in the missing values, by computing the priors. (Just use a piece of paper, you will need them for a later question)

<div align="center">Temperature

|| Yes        | No | P(yes)|P(no)
| ------------- |-------------| ------|--------|----|
| Hot   | 3| 1 |3/5|1/5|
| Cold    | 2      | ?|P(Cold \| yes)=?|P(Cold \| no)=?|
| Total   |5| 5 |100%|100%|
</div>

<div align="center">Weather

|| Yes        | No | P(yes)|P(no)
| ------------- |-------------| ------|--------|----|
| Sunny   | 2| 3 |2/5|3/5|
| Rainy   | ?     |  ?|P(rainy \| yes)=?|P(Rainy \| no)=?|
| Total   |5| 5 |100%|100%|
</div>

<div align="center">Humidity

|| Yes        | No | P(yes)|P(no)
| ------------- |-------------| ------|--------|----|
| High   | 4| 1 |4/5|1/5|
| Low  | ?     |  ?|?|?|
| Total   |5| 5 |100%|100%|
</div>

## Question 1

(2 pts.) So, in the figures above, we have calculated $P(x_i | y_j)$ for each $x_i$ in $X$ and $y_j$ in $y $ manually in the tables 1-4. For example, probability of playing golf given that the temperature is cool, i.e P(temp. = cold | play golf = Yes) = 2/5.

Also, we need to find class probabilities $(P(y))$. For example, P(play golf = Yes) = 5/10.
<div align="center">Play Golf 
    

| Play     | P(yes)|P(no)
| ------------- | ------|--------|
| Yes |  5|?|
|No | 5|5/10|
| Total   |10| 100%|
</div>

today = (Sunny, Hot, Low)

so $P(Yes|today)$ = $P(Sunny | Yes)* P(Hot | Yes) * P(Low | Yes) * P(Yes)$ (we can ignore denominator)
= ?

<b style="color:red">Type in the four fractions below, and the result when you multiply them together.</b>

P(Sunny|Yes) = 2/5, From 1 - P(Sunny|No) <br> 
P(Hot|Yes) = 3/5; From 1 - P(Hot|No) <br>
P(Low|Yes) = 1/5; From 1 - P(Low|No) <br>
P(Yes) = 5/10; From 1 - P(No) <br>
P(Yes|Today) = 3/125

## Question 2

(1 pt.) Now, what is P(No | Today)?
Next, convert these two values into probabilities by making the sum equal to 1 (normalization):  example: P(Yes|today)/ P(Yes |today) + P(No | today)
<b style="color:red">Write the probability that golf will be played today.
How likely is it golf will be played today?</b>

P(No|Today) = P(Sunny|No) * P(Hot|No) * P(Low|No) * P(No) <br>
P(Sunny|No) = 3/5 <br>
P(Hot|No) = 1/5 <br>
P(Low|No) = 4/5 <br>
P(No) = 5/10 <br>
P(No|Today) = 6/125 <br>

Normalization: <br>
P(Yes|Today) = 3/125 / (3/125 + 6/125) = 1/3 <br>
P(No|Today) = 1 - P(Yes|Today) = 2/3

# Iris Classification with Naive Bayes

In this part of the assignment, you will again use the [Iris dataset](https://scikit-learn.org/stable/auto_examples/datasets/plot_iris_dataset.html). Recall, it consists of 50 samples from each of three species of Iris (Iris setosa, Iris virginica and Iris versicolor). Four features were measured from each sample: the length and the width of the sepals and petals, in centimeters. For a reference, see the following papers:

- R. A. Fisher. "The use of multiple measurements in taxonomic problems". Annals of Eugenics. 7 (2): 179–188, 1936.

Your goal is to construct a Naive Bayes classifier model that predicts the correct class from the sepal length and sepal width features. This lab will help learn about using probability distributions in the Naive Bayes classifier. 

In [1]:
# run this cell
#### PACKAGE IMPORTS ####
import numpy as np
import matplotlib.pyplot as plt
import sklearn as sk
from sklearn.metrics import accuracy_score
from sklearn import datasets, model_selection
import pandas as pd
import seaborn as sns
%matplotlib inline

# Step 1: Look at the big picture

What is the problem you want to solve? How do you plan to use and benefit from the the machine learning model? What is your metric for success?

In this exercise, we will give you the goal and the metric:

**Machine Learning Goal**: classify irises based on the measurement of their petals. 

**Metric**: accuracy - defined as the number of correctly typed irises divided by the total number of irises


# Step 2: Get the data

What data are available to you for training your model? How will you access and download all the data? The type and quantity of data available will often impact your choice in the subsequent stages. 

Scikit Learn has a bunch of pre-built machine learning datasets that are helpful for learning new models and testing. 
#### Load and prepare the data

We will first read in the Iris dataset, and then split the dataset into training and test sets. The extra cell below the code solution is provided to check your work.

## Question 3


In [ ]:
# (1 pt.) Load the iris dataset (hint: see how Lab 1 loaded the Iris dataset)
# YOUR CODE HERE
iris = datasets.load_iris()
df = pd.DataFrame(data = np.c_[iris['data'], iris['target']], columns = iris['feature_names'] + ['target'])
print(df)

# Step 3: Explore the data

Discover and visualize the data to gain insight. For example, are there more representatives of certain types of data than others?


## Question 4
(6 pts)

For this lab, please answer the following questions about your data. Use as many cells and plots as you like. 

<b style="color:red">1. What are the features? What is the label?   
2. Compute some summary statistics about your dataset. For example, the number of samples, mean, median, dispersion of the features.   
3. Plot some of the features to get a sense of the data.</b>

1. The features are the petal length, petal width, sepal length, and sepal width. The features are measured in centimeters (cm). The label is the iris, which is enumerated.

In [ ]:
print(df.mean())

In [ ]:
print(df.median())

In [ ]:
print(df.quantile(0.25))

In [ ]:
print(df.quantile(0.75))

In [ ]:
print(df.std())

In [ ]:
print(df.var())

In [ ]:
ax = plt.axes()
features = df.drop("target", axis = 1).select_dtypes('number').dropna()
sns.heatmap(features.corr(), annot = True, ax = ax)
ax.set_title("Iris Feature Correlation")

# Step 4: Prepare the data for ML algorithms

Machine learning algorithms often require data to be pre-processed in a certain way, such as scaling numerical values or map categories to other representations. You will also need to decide what data to use as training and as testing. You may also need to develop a strategy to deal with missing data. 

Use only the first two features: **sepal length** and **width** for this assignment

## Question 5

In [ ]:
# (1 pt.) Create an array called data with the sepal length and width features and
# Create an array called targets with the target values
# YOUR CODE HERE

f_data = df[['sepal length (cm)', 'sepal width (cm)']]
t_data = df['target'].astype('int')
data = f_data.to_numpy()
targets = t_data.to_numpy()
#raise NotImplementedError()

`sklearn` has a really helpful function called `train_test_split` that helps to randomly split your data and targets for training and testing. Please read the documentation on `train_test_split` before running the following cell. Note that there is an important keyword `test_size`. (Look at what random_seed can do for you as well, we will use that later.)

A short summary of terminology used below:

* **training** - data used in training the machine learning model
* **test** - data used to evaluate the performance of the model. Will use this information to further refine the model. 
* **validation** - the data left aside in the beginning only to be used at the very end when you are finished with your model and want to test on data the model has never seeen. 

In [ ]:
# Run this code to randomly shuffle the data and make train and validation splits

x_train_all, x_validate, y_train_all, y_validate = model_selection.train_test_split(data, targets, test_size=0.1)

# now split the training data further into training and testing
x_train, x_test, y_train, y_test = model_selection.train_test_split(x_train_all, y_train_all, test_size=0.2)


In [ ]:
# Example of plotting the training data

labels = {0: 'Iris-Setosa', 1: 'Iris-Versicolour', 2: 'Iris-Virginica'}
label_colours = ['blue', 'orange', 'green']

def plot_data(x, y, labels, colours):
    for c in np.unique(y):
        inx = np.where(y == c)
        plt.scatter(x[inx, 0], x[inx, 1], label=labels[c], c=colours[c])
    plt.title("Training set")
    plt.xlabel("Sepal length (cm)")
    plt.ylabel("Sepal width (cm)")
    plt.legend()
    
plt.figure(figsize=(8, 5))
plot_data(x_train, y_train, labels, label_colours)

# Step 5: Select a model and train it

Based on the problem and the data, there are often a handful of algorithms to try. Here, experimentation and knowledge of the strengths and weaknesses of machine learning models will help you choose a model and train it.

In this lab, we will specifically use the Naive Bayes model. 
### Naive Bayes classifier

We will briefly review the Naive Bayes classifier model. The fundamental equation for this classifier is Bayes' rule:

$$
P(Y=y_k | X_1,\ldots,X_d) = \frac{P(X_1,\ldots,X_d | Y=y_k)P(Y=y_k)}{\sum_{k=1}^K P(X_1,\ldots,X_d | Y=y_k)P(Y=y_k)}
$$

In the above, $d$ is the number of features or dimensions in the inputs $X$ (in our case $d=2$), and $K$ is the number of classes (in our case $K=3$). The distribution $P(Y)$ is the class prior distribution, which is a discrete distribution over $K$ classes. The distribution $P(X | Y)$ is the class-conditional distribution over inputs.

The Naive Bayes classifier makes the assumption that the data features $X_i$ are conditionally independent give the class $Y$ (the 'naive' assumption). In this case, the class-conditional distribution decomposes as

$$
\begin{align}
P(X | Y=y_k) &= P(X_1,\ldots,X_d | Y=y_k)\\
&= \prod_{i=1}^d P(X_i | Y=y_k)
\end{align}
$$

This simplifying assumption means that we typically need to estimate far fewer parameters for each of the distributions $P(X_i | Y=y_k)$ instead of the full joint distribution $P(X | Y=y_k)$.

Once the class prior distribution and class-conditional densities are estimated, the Naive Bayes classifier model can then make a class prediction $\hat{Y}$ for a new data input $\tilde{X} := (\tilde{X}_1,\ldots,\tilde{X}_d)$ according to

$$
\begin{align}
\hat{Y} &= \text{argmax}_{y_k} P(Y=y_k | \tilde{X}_1,\ldots,\tilde{X}_d) \\
&= \text{argmax}_{y_k}\frac{P(\tilde{X}_1,\ldots,\tilde{X}_d | Y=y_k)P(Y=y_k)}{\sum_{k=1}^K P(\tilde{X}_1,\ldots,\tilde{X}_d | Y=y_k)P(Y=y_k)}\\
&= \text{argmax}_{y_k} P(\tilde{X}_1,\ldots,\tilde{X}_d | Y=y_k)P(Y=y_k)
\end{align}
$$

For this lab, we will use **Gaussian distributions** to describe the distribution of sepal length and width for each of the classes of irises. 

## 5a Compute the Priors
First, let's compute the priors for observing either three iris species. Without more information, we'll just use the fraction of each classes in the training data as the prior on that class. 

In [ ]:
counter = {0 : 0, 1 : 0, 2 : 0} # Sets up a dictionary to keep track of relevant classifications
tot = 0 # Normalization Const
for x in y_train: # run through training set to find priors
    counter[x] += 1 
    tot += 1
for x in counter: # Lets me know what the priors are
    print(f"class {x} ({labels[x]}) has a fraction of {counter[x]/tot}")

## Question 6

In [ ]:
# (1 pt.) Create an array called priors that has the prior probability of observing each class.
# Be sure to keep track of which class belongs to which element of the array!
# YOUR CODE HERE
priors = []
for x in counter:
    priors.append(counter[x]/tot) #track priors
    #raise NotImplementedError()

## 5b Build the Naive Bayes Classifier



## Question 7

<b style="color:red">Build the Naive Bayes Classifier using Gaussian distributions. </b>

In [2]:
# (3 pt.) Build Classifier here
# YOUR CODE HERE
def g_parameters(data, tgt, feature, target): #General function to get the mean and variance of a specific feature and tgt from the testing data
    focused_set = data.loc[data[target] == tgt] # Creates relevant dataset with desired tgt/feature
    mean = focused_set[feature].mean()
    var = focused_set[feature].var()
    return mean, var

def gaussian(mean, var, inp): # Gaussian function
    return 1/(np.sqrt(var * 2 * np.pi)) * np.e**(-(inp - mean)**2/(2 * var))


In [ ]:
#----Creates a combined data set to run through my parameters function----
targets = pd.DataFrame(pd.Series(y_train), columns = ['target']) 
features = pd.DataFrame(x_train, columns = ['sepal length (cm)', 'sepal width (cm)'])
combined = pd.concat([targets, features], axis = 1)

#----All 12 relevant parameters (2 features * 3 classes * 2 parameters required per set)----
g_l0_mean, g_l0_var = g_parameters(combined, 0, "sepal length (cm)", 'target')
g_l1_mean, g_l1_var = g_parameters(combined, 1, "sepal length (cm)", 'target')
g_l2_mean, g_l2_var = g_parameters(combined, 2, "sepal length (cm)", 'target')
g_w0_mean, g_w0_var = g_parameters(combined, 0, "sepal width (cm)", 'target')
g_w1_mean, g_w1_var = g_parameters(combined, 1, "sepal width (cm)", 'target')
g_w2_mean, g_w2_var = g_parameters(combined, 2, "sepal width (cm)", 'target')

## 5c Train the model

Training the model in this case is to determine the parameters of the Gaussians describing the joint probability of observing the sepal length or width with respect to the class. 


## Question 8

(3 pt.)<b style="color:red"> Using as many cells as you need, determine the parameters for the joint probability distribution.</b>

In [ ]:
#----Combine the parameters into an array for easy/general access----
parameters = [[g_l0_mean, g_l0_var, g_w0_mean, g_w0_var], [g_l1_mean, g_l1_var, g_w1_mean, g_w1_var], [g_l2_mean, g_l2_var, g_w2_mean, g_w2_var]]

## 5d Test the model predictions

`sklearn` comes with a lot of useful routines to compute metrics for machine learning applications. In this lab, you'll use `accuracy_score` from `sklearn.metrics` to compute the model accuracy.

## Question 9

In [ ]:
# (1 pt.) make a array called pred that is your prediction for most likely class given your test data
# YOUR CODE HERE
pred = []
curr = -1
curr_tgt = -1

#----Compares probability score from gaussian distribution naive bayes model to find most likely target----
for x in x_test:
    for y in range (0, 3):
        tmp = priors[y] * gaussian(parameters[y][0], parameters[y][1], x[0]) * gaussian(parameters[y][2], parameters[y][3], x[1])
        if tmp > curr: # Updates max score and most likely target
            curr_tgt = y
            curr = tmp
    pred.append(curr_tgt) # Adds prediction, then resets model for the next iteration
    curr = -1
    curr_tgt = -1
print(pred)
#raise NotImplementedError()

In [ ]:
# this line will give you your accuracy score for the test data
score = accuracy_score(y_test, pred)
print(score)

## Question 10

For this case, we will not modify our model, so we can just go ahead evaluate the accuracy of the validation data that you saved earlier. 

<b style="color:red">Below, find the accuracy of the validation data. </b>

In [ ]:
# (1pt.) determine the accuracy score of your validation data
validate = []
curr = -1
curr_tgt = -1
for x in x_validate:
    for y in range (0, 3):
        tmp = priors[y] * gaussian(parameters[y][0], parameters[y][1], x[0]) * gaussian(parameters[y][2], parameters[y][3], x[1])
        if tmp > curr:
            curr_tgt = y
            curr = tmp
    validate.append(curr_tgt)
    curr = -1
    curr_tgt = -1
score = accuracy_score(y_validate, validate)
print(score)

Congratulations! You have now completed the iris classification. In this iris classification, we won't go through Steps 6 to 8 of the machine learning workflow, but we will in the next part of this lab. 

# Classifying Stars at the Galactic Center

The Galactic center contains a complex population of stars. Multiple generations of star formation have occured, including as recently as 4 to 6 million years ago. The most reliable way to determine the ages of the stars is measure their spectra to examine their spectral features. Generally stars near the supermassive black hole in Galactic center are old (> 1 billion years) or young (4-6 million years). The stars that are bright enough for us to see are young stars and massive (> 10 Msun) with high surface temperatures (>20,000K), or old stars in their red giant phase with lower mass (~1 Msun) and cool temperatures (3000 to 4000 K). This difference in temperature leads to very different spectral features. 

In the spectra below, there are two examples of young stars and one example of an old star. The young star spectra are classified as 'early-type'. Because the temperature is high, many electrons have been ionized from atoms in its atmosphere so few absorption lines remain. The most prominent absorption line for early-type stars in this wavelength region is from hydrogen (specifically, the Brackett gamma line). Sometimes, the temperature is so high that even the electron in hydrogen is ionized, so there is very little hydrogen absorption. For the old red-giant stars, the temperatures are fairly low, so there are now many atomic absorption lines. The strongest lines in this wavelength range are from sodium (Na) atoms. In addition, because the temperature is cool, the hydrogen lines become less prominant because there is not enough high energy photons that are available to make those absorption lines. 

![gc_spectra](gc_spectra.png)

## Feature Engineering: Equivalent Widths and Cross-Correlation

To simplify and automate the process of classifying stars, we can reduce the spectrum (flux measurements at different wavelengths) into a more compact representation that hopefully still carries a lot of the information necessary for classification. In this case, we pick the strongest spectral lines (Br gamma and Na) and turn them into features for our machine learning models. We will use the concept of equilvalent width, which is related to the depth of the absorption features. In the figure above, the shaded blue regions represent the equivalent widths of the lines. The stronger the lines, the large the equivalent widths. For more details see: https://en.wikipedia.org/wiki/Equivalent_width. We will define positive equivalent widths as absorption lines (like in the figure above), and negative equivalent widths as emission lines. 

We will also use the cross-correlation coefficient of the spectra with templates as a potential feature. This feature is the maximum from cross-correlating a template young star and old star spectrum with the observed spectrum. You can think of this correlation coefficient as how closely matched the observed spectrum is with a template spectrum. The correlation in principle should go from 0 to 1. With 0 meaning it is not at all alike and 1 being an identical match. The correlation coefficent given in the data for this lab varies widely because the observed spectra often have more noise the the spectrum above and we only cross correlate the observed spectra with one model young star and one model old star. There are also intrinsic variations in the spectra that can also reduce the correlations. 

The dataset you'll be using has the equivalent widths and cross correlation coefficient measured using an automated method for spectra that have first been classified by humans, so there are labels. 

# Step 1: Look at the big picture

What is the problem you want to solve? How do you plan to use and benefit from the the machine learning model? What is your metric for success?

In this lab, we will give you the goal and the metric

**Lab goal**: classify stars at the Galactic center into young or old. 

**Metric**: accuracy - defined as the number of correctly typed stars divided by the total number of stars


# Step 2: Get the data

What data are available to you for training your model? How will you access and download all the data? The type and quantity of data available will often impact your choice in the subsequent stages. 

Here, the data is straightforward. We have provided you with the training data in the file ``galactic_center_stars_training.csv``

In the training data, there should be 8 columns: Na equivalent width, Na equivalent width error, correlation coefficient with an old star template, Br gamma equivalent width, Br gamma equivalent width uncertainty, signal-to-noise ratio of the spectrum, and young or old. The units of equilvalent widths and their uncertainties are in Angstroms. The correlation coefficients are unit-less. 

There is also a file called ``galactic_center_stars_eval.csv`` that has no labels. We'll be using that file to grade this assignment. 


## Question 11

In [209]:
# (1 pt.) In this cell, load the training data and the data for grading into two pandas data frames. 
# and check that they are there by looking at each head.
# YOUR CODE HERE
path = "galactic_center_stars_training.csv"
df = pd.read_csv(path).dropna() # We do not want to be training on NaN data


df.head
#df_eval.head()

#raise NotImplementedError()


<bound method NDFrame.head of        Na  Na_err   Corr    Br  Br_err  Yng_corr    SNR   Type
0    1.11    0.91   0.81  2.57    0.53      0.54  17.69    old
1    5.40    0.78   0.77  0.98    0.26      0.36  20.74    old
2    5.71    0.80   0.81  0.96    0.22      0.34  20.23    old
3    2.54    1.45   0.93  2.94    0.87      0.11  10.92    old
4    7.13    1.42   0.83  1.33    0.46      6.28  11.13    old
..    ...     ...    ...   ...     ...       ...    ...    ...
414  0.16    0.59   0.44  1.26    0.25      0.48  27.61  young
415  1.78    0.68   0.74  1.01    0.22      0.49  23.98  young
416 -0.14    2.11  33.54  7.66    1.57      0.42   7.38  young
417 -0.14    1.14  30.93  3.47    0.59      0.64  13.94  young
418  0.74    1.04   0.18  3.26    0.87      0.42  15.39  young

[417 rows x 8 columns]>

# Step 3: Explore the data

Discover and visualize the data to gain insight. For example, how complete is the dataset? Are there more representatives of certain types of data than others?


## Question 12
(6 pts)

For this lab, please answer the following questions about your data. Use as many cells and plots as you like. 

<b style="color:red">1. What are the features? What is the label?   
2. Compute some summary statistics about your dataset. For example, the number of samples, mean, median, dispersion of the features.   
3. Plot some of the features to get a sense of the data. </b>

1) The features are the Na and Br gamma spectral line widths and the label is the age descriptor of the star (old or young)

In [ ]:
print(f" The number of samples in the training data set is {len(df)}")

In [ ]:
# use as many cells as you need to explore the data. We will be grading manually


In [ ]:
print(f" The mean of the Na spectral line width is {df['Na'].mean()}")
print(f" The median of the Na spectral line width is {df['Na'].median()}")
print(f" The first quantile of the Na spectral line width is {df['Na'].quantile(0.25)}")
print(f" The thrid quantile of the Na spectral line width is {df['Na'].quantile(0.75)}")
print(f" The variance of the Na spectral line width is {df['Na'].var()}")

In [ ]:
print(f" The mean of the Br spectral line width is {df['Br'].mean()}")
print(f" The median of the Br spectral line width is {df['Br'].median()}")
print(f" The first quantile of the Br spectral line width is {df['Br'].quantile(0.25)}")
print(f" The thrid quantile of the Br spectral line width is {df['Br'].quantile(0.75)}")
print(f" The variance of the Br spectral line width is {df['Br'].var()}")

In [ ]:
# You can try sns.pairplot(train_tab, hue="Type")
sns.pairplot(df, hue = "Type")

# Step 4: Prepare the data for ML algorithms

Machine learning algorithms often require data to be pre-processed in a certain way, such as scaling numerical values or map categories to other representations. You will also need to decide what data to use as training and as testing. You may also need to develop a strategy to deal with missing data. 

For this lab, we have already done the feature engineering to get the equivalent widths, their uncertainties, and the correlation coefficients. 

## Question 12
(6 pts)

While the features are already chosen, there are a number of other questions that need to be answered:

<b style="color:red">1. Are there bad data points or outliers?   
2. If there are outliers, what is your strategy for dealing with them?   
3. Apply your strategy for cleaning and dealing with the data, then split your training data into three parts: training, testing, and validation  </b>


1) Yes, there are bad data points in the dataset. For example, there are data points where the Br emission lines are negative or in the tens of thousands.
2) Due to the large mean value of the Br emission line, I plan on using interquartile ranges to eliminate values. To clean the data, I will remove any data point beyond 1.5 * IQR past the first and third quartiles, as per the IQR rule.

In [232]:
df_new = df
for feat in df.select_dtypes(include = 'number').columns:
    Q1 = df[feat].quantile(0.25)
    Q3 = df[feat].quantile(0.75)
    IQR = Q3 - Q1
    low = Q1 - 1.5 * IQR
    high = Q3 + 1.5 *IQR
    df_new = df_new[(df_new[feat] >= low) & (df_new[feat] <= high)]
#df_new = df.drop(outlier.index)

df_new['Type'] = df_new['Type'].apply(lambda x: 0 if x == 'young' else 1)
df_new.head

<bound method NDFrame.head of        Na  Na_err  Corr    Br  Br_err  Yng_corr    SNR  Type
0    1.11    0.91  0.81  2.57    0.53      0.54  17.69     1
1    5.40    0.78  0.77  0.98    0.26      0.36  20.74     1
2    5.71    0.80  0.81  0.96    0.22      0.34  20.23     1
3    2.54    1.45  0.93  2.94    0.87      0.11  10.92     1
5    5.11    0.71  0.85  0.79    0.18      0.28  22.79     1
..    ...     ...   ...   ...     ...       ...    ...   ...
407  1.06    0.81  0.72 -1.06    0.27      0.18  19.93     0
413  0.70    0.73  0.53  3.68    0.49      0.64  22.15     0
414  0.16    0.59  0.44  1.26    0.25      0.48  27.61     0
415  1.78    0.68  0.74  1.01    0.22      0.49  23.98     0
418  0.74    1.04  0.18  3.26    0.87      0.42  15.39     0

[257 rows x 8 columns]>

In [233]:
# Describe your process, clean the data, and show clean_training.describe()
"Data was cleaned using the 1.5 IQR Rule to elimiante outliers. Outliers 1.5 * IQR above the 3rd quartile and 1.5 * IRQ below the first quartile in any column"  
"besides type were removed from the data set."
df_new.describe()

Na      Na_err        Corr          Br      Br_err    Yng_corr  \
count  257.000000  257.000000  257.000000  257.000000  257.000000  257.000000   
mean     3.465253    0.929105    0.699728    1.984086    0.519844    0.418872   
std      1.905164    0.273731    0.154812    1.765167    0.320190    0.225118   
min     -1.050000    0.440000    0.180000   -3.050000    0.140000   -0.200000   
25%      2.230000    0.730000    0.590000    0.960000    0.260000    0.270000   
50%      3.830000    0.900000    0.720000    1.640000    0.460000    0.380000   
75%      4.900000    1.070000    0.830000    3.040000    0.670000    0.540000   
max      7.440000    1.750000    0.940000    7.330000    1.670000    1.240000   

              SNR        Type  
count  257.000000  257.000000  
mean    18.950156    0.817121  
std      5.915087    0.387322  
min      8.920000    0.000000  
25%     14.920000    1.000000  
50%     17.800000    1.000000  
75%     22.200000    1.000000  
max     37.650000    1.000000

In [ ]:
# Before splitting the training data into trainig, testing, and validation sets,
# plot clean training with sns.pairplot(clean_training, hue="Type")
sns.pairplot(df, hue="Type")
plt.show()

In [211]:
# After cleaning, split data into training, testing, and validation
data = df_new.drop('Type', axis = 1)
x_train_all, x_validate, y_train_all, y_validate = model_selection.train_test_split(data, df_new['Type'], test_size=0.1)
x_train, x_test, y_train, y_test = model_selection.train_test_split(x_train_all, y_train_all, test_size=0.2)
# YOUR CODE HERE
#raise NotImplementedError()

# Step 5: Select a model and train it

Based on the problem and the data, there are often a handful of algorithms to try. Here, experimentation and knowledge of the strengths and weaknesses of machine learning models will help you choose a model and train it.


## Question 13
(10 pts)

In this lab, we will specifically use the Naive Bayes model. 

In the following cells, please implement the Naive Bayes classifier to deterimine whether a star is young or old based on the features you've selected in Step 4.  
<b style="color:red">At the end of this step, report the accuracy of your model on the test data.</b> 

In [128]:
# use as many cells as you need to do Step 5. We will be grading manually
features = x_train[['Na', 'Br']]
tgt = pd.DataFrame({'Type': y_train})
galactic_combined = pd.concat([tgt, features], axis = 1)

p1 = sum(y_train)
tot = len(y_train)

priors = [1 - (p1 / tot), (p1 / tot)]
print(priors)

g_n0_mean, g_n0_var = g_parameters(galactic_combined, 0, "Na", 'Type')
g_n1_mean, g_n1_var = g_parameters(galactic_combined, 1, "Na", 'Type')
g_b0_mean, g_b0_var = g_parameters(galactic_combined, 0, "Br", 'Type')
g_b1_mean, g_b1_var = g_parameters(galactic_combined, 1, "Br", 'Type')
parameters = [[g_n0_mean, g_n0_var, g_b0_mean, g_b0_var], [g_n1_mean, g_n1_var, g_b1_mean, g_b1_var]]

pred = []
labels = ['young', 'old']

for x in x_test.to_numpy():
    curr = -1
    curr_tgt = -1
    for y in range (0, 2):
        tmp = priors[y] * gaussian(parameters[y][0], parameters[y][1], x[0]) * gaussian(parameters[y][2], parameters[y][3], x[1])
        if tmp > curr: # Updates max score and most likely target
            curr_tgt = y
            curr = tmp
    pred.append(curr_tgt) # Adds prediction, then resets model for the next iteration

[0.18478260869565222, 0.8152173913043478]


In [129]:
# Uncomment these and run when ready
from sklearn.metrics import accuracy_score
accuracy_score(y_test, pred)

0.9574468085106383

# Step 6: Fine tune the model

Most machine learning algorithms will have hyperparameters that can be tuned to improve the performance of the model given the specific dataset. You may also want to visualize the results of model predictions to aid in tuning the model.


## Question 14
(9 pts)

For this lab, please do the following:  
<b style="color:red">
1. One of the hyperparameters in this case is the types of features to use. Try subtracting or adding a feature to see how it changes your accuracy.   
2. Examine how changes in your training data might affect your results. For example, what if you use only measurements with small uncertainties for training?   
3. Report your final accuracy on your test dataset after selecting the optimal model. Explain why you think this is the optimal model.</b> 

In [187]:
tests = 1000 # Make as many runs as possible to make an average guess on accuracy

In [212]:
# use as many cells as you need to do Step 6. We will be grading manually
# 3 Feature Model (Na, Br emission lines, and Na Correlation)
avg = 0

features = x_train[['Na', 'Br', 'Corr']] # Takes features we care about
tgt = pd.DataFrame({'Type': y_train}) 
galactic_combined = pd.concat([tgt, features], axis = 1)
    
p1 = sum(y_train) # Due to precomputing all old == 1 and young == 0, we can simply sum the y_train dataset to find the number of old classifications
tot = len(y_train) # Total
    
priors = [1 - (p1 / tot), (p1 / tot)] # Priors in terms of percentages

#---Gaussian Parameter Calculations---
params = [] 
for y in range(0, 2): # 2 classes -> range (0, 2)
    curr = []
    for x in galactic_combined: # runs over our dataframe with the notable features and types, we are running over every feature
        if x == 'Type': # we do not want to compute gaussian parameters for the type
            continue
        mean, var = g_parameters(galactic_combined, y, x, 'Type') # Finds mean and var by running through the parameters function 
        curr.append(mean) 
        curr.append(var)
    params.append(curr) # [[mean_feature1_class1, var_feature1_class1, mean_feature2_class1, ...], [mean_feature1_class2, ...], ...]
        
for i in range(0, tests):
    x_train, x_test, y_train, y_test = model_selection.train_test_split(x_train_all, y_train_all, test_size=0.2) # Obtain a new set of testing data
    
    pred = []
    
    for x in range (0, len(x_test)): # Find the maximum of the argumet: prior * gaussian distribution of feature 1 * gaussian distribution of feature 2 * ...
        curr = -1
        curr_tgt = -1
        for y in range (0, 2):
            tmp = priors[y] * gaussian(params[y][0], params[y][1], x_test.iloc[x].Na) * gaussian(params[y][2], params[y][3], x_test.iloc[x].Br) * gaussian(params[y][4], params[y][5], x_test.iloc[x].Corr)
            if tmp > curr: # Updates max score and most likely target
                curr_tgt = y
                curr = tmp
        pred.append(curr_tgt) # Adds prediction, then resets model for the next iteration
    avg += accuracy_score(y_test, pred) # add accuracy ([0, 1]) to avg
print(avg / tests) # divide by total number of tests to find average

0.9697234042553248


In [213]:
# 2 Feature Model (Na, Br emission lines)
avg = 0
features = x_train[['Na', 'Br']]
tgt = pd.DataFrame({'Type': y_train})
galactic_combined = pd.concat([tgt, features], axis = 1)
    
p1 = sum(y_train)
tot = len(y_train)
    
priors = [1 - (p1 / tot), (p1 / tot)]
    
params = []
for y in range(0, 2):
    curr = []
    for x in galactic_combined:
        if x == 'Type':
            continue
        mean, var = g_parameters(galactic_combined, y, x, 'Type')
        curr.append(mean)
        curr.append(var)
    params.append(curr)
    
for i in range (0, tests):
    x_train, x_test, y_train, y_test = model_selection.train_test_split(x_train_all, y_train_all, test_size=0.2)
    
    pred = []
    labels = ['young', 'old']
    
    for x in range (0, len(x_test)):
        curr = -1
        curr_tgt = -1
        for y in range (0, 2):
            tmp = priors[y] * gaussian(params[y][0], params[y][1], x_test.iloc[x].Na) * gaussian(params[y][2], params[y][3], x_test.iloc[x].Br)
            if tmp > curr: # Updates max score and most likely target
                curr_tgt = y
                curr = tmp
        pred.append(curr_tgt) # Adds prediction, then resets model for the next iteration
    avg += accuracy_score(y_test, pred)
print(avg / tests)

0.9745319148936225


In [226]:
# Small Error Model
avg = 0

#---Dataframe prepation--- 
df_err = df # Working with original dataset
remove = pd.DataFrame()
for i in range (0, len(df)): # For the features we are using, remove any error > 37%
    if (df_err.iloc[i].Na != 0 and abs(df_err.iloc[i].Na_err / df_err.iloc[i].Na) > 0.37): 
        remove = pd.concat([remove, df_err.iloc[i].to_frame().T])
for i in range (0, len(df)):
    if (df_err.iloc[i].Br != 0 and abs(df_err.iloc[i].Br_err / df_err.iloc[i].Br) > 0.37): 
        remove = pd.concat([remove, df_err.iloc[i].to_frame().T])
        
df_err = df_err.drop(remove.index)
df_err['Type'] = df_err['Type'].apply(lambda x: 0 if x == 'young' else 1)

data = df_err.drop('Type', axis = 1)

x_train_all, x_validate, y_train_all, y_validate = model_selection.train_test_split(data, df_err['Type'], test_size=0.1)
x_train, x_test, y_train, y_test = model_selection.train_test_split(x_train_all, y_train_all, test_size=0.2)

#---Prior Calculation---
features = x_train[['Na', 'Br']]
tgt = pd.DataFrame({'Type': y_train})
galactic_combined = pd.concat([tgt, features], axis = 1)
    
p1 = sum(y_train)
tot = len(y_train)
    
priors = [1 - (p1 / tot), (p1 / tot)]
params = []

#---Gaussian Parameter Calculation---
for y in range(0, 2):
    curr = []
    for x in galactic_combined:
        if x == 'Type':
            continue
        mean, var = g_parameters(galactic_combined, y, x, 'Type')
        curr.append(mean)
        curr.append(var)
    params.append(curr)

#---Testing---
for i in range (0, tests):
    x_train, x_test, y_train, y_test = model_selection.train_test_split(x_train_all, y_train_all, test_size=0.2)
    
    pred = []

    for x in range(0, len(x_test)):
        curr = -1
        curr_tgt = -1
        for y in range (0, 2):
            tmp = priors[y] * gaussian(params[y][0], params[y][1], x_test.iloc[x].Na) * gaussian(params[y][2], params[y][3], x_test.iloc[x].Br)
            if tmp > curr: # Updates max score and most likely target
                curr_tgt = y
                curr = tmp
        pred.append(curr_tgt) # Adds prediction, then resets model for the next iteration
    avg += accuracy_score(y_test, pred)
print(avg / tests)

0.9941212121212158


1) Generally, adding more features seems to decrease the accuracy of the model
2) Small Error model was created above
3) Best Score: Average of 99.41% correctness from Small Error (>37% error eliminated) based on 1000 total tests

# Step 7: Present your solution and apply it

Here you will evaluate your model against the test set that was set aside in Step 2 to determine the final performance metrics. Your model is now ready to be applied to new data.

## Question 15
(2 pts)

For this lab, please do the following:  
<b style="color:red">
1. Report your accuracy for the validation data that you saved in Step 4. How does this score compare to when you were training and fine tuning your model?  
2. Apply your model to the data for the unlabeled grading data. Note that you should explore these inputs to see if there are differences in the features between your training data and this dataset. Beware of outliers in this dataset as well. Create a set of predictions called pred_eval.</b> 

In [229]:
# use as many cells as you need to do Step 7. 
df_err = df
remove = pd.DataFrame()
for i in range (0, len(df)):
    if (df_err.iloc[i].Na != 0 and abs(df_err.iloc[i].Na_err / df_err.iloc[i].Na) > 0.37):
        remove = pd.concat([remove, df_err.iloc[i].to_frame().T])
for i in range (0, len(df)):
    if (df_err.iloc[i].Br != 0 and abs(df_err.iloc[i].Br_err / df_err.iloc[i].Br) > 0.37): 
        remove = pd.concat([remove, df_err.iloc[i].to_frame().T])

df_err = df_err.drop(remove.index)
df_err['Type'] = df_err['Type'].apply(lambda x: 0 if x == 'young' else 1)

data = df_err.drop('Type', axis = 1)

x_train_all, x_validate, y_train_all, y_validate = model_selection.train_test_split(data, df_err['Type'], test_size=0.1)
x_train, x_test, y_train, y_test = model_selection.train_test_split(x_train_all, y_train_all, test_size=0.2)

features = x_train[['Na', 'Br']]
tgt = pd.DataFrame({'Type': y_train})
galactic_combined = pd.concat([tgt, features], axis = 1)
    
p1 = sum(y_train)
tot = len(y_train)
    
priors = [1 - (p1 / tot), (p1 / tot)]
params = []
for y in range(0, 2):
    curr = []
    for x in galactic_combined:
        if x == 'Type':
            continue
        mean, var = g_parameters(galactic_combined, y, x, 'Type')
        curr.append(mean)
        curr.append(var)
    params.append(curr)
    
pred = []
labels = ['young', 'old']
    
for x in range(0, len(x_validate)):
    curr = -1
    curr_tgt = -1
    for y in range (0, 2):
        tmp = priors[y] * gaussian(params[y][0], params[y][1], x_validate.iloc[x].Na) * gaussian(params[y][2], params[y][3], x_validate.iloc[x].Br)
        if tmp > curr: # Updates max score and most likely target
            curr_tgt = y
            curr = tmp
    pred.append(curr_tgt) # Adds prediction, then resets model for the next iteration
accuracy_score(y_validate, pred)

0.9473684210526315

If you would like to submit your group's solution to the unlabeled dataset for the competition to see who gets the accurate predictions, please send your TA and Prof. Do a csv file with a single column predicting whether each star is 'young' or 'old'. Below is an example code of how to create this file. You will send ONE file per team. 

In [230]:
import pandas as pd

# if your answer for three stars were:
#answer = ['old','young','old']


path_comp = "galactic_center_stars_eval.csv"
df_comp = pd.read_csv(path_comp)

for x in range (0, len(df_comp)):
    curr = -1
    curr_tgt = -1
    for y in range (0, 2):
        tmp = priors[y] * gaussian(params[y][0], params[y][1], df_comp.iloc[x].Na) * gaussian(params[y][2], params[y][3], df_comp.iloc[x].Br)
        if tmp > curr: # Updates max score and most likely target
            curr_tgt = y
            curr = tmp
    pred.append(curr_tgt) # Adds prediction, then resets model for the next iteration

answer = []
for x in pred:
    if x == 0:
        answer.append('young')
    else:
        answer.append('old')

# change the list into a data frame
answer_tab = pd.DataFrame({'target':answer})

# save the data frame into a csv file
answer_tab.to_csv('group_x_answer.csv',index=None)